modules/seqexec/server/src/main/scala/seqexec/server/gnirs/GnirsControllerEpics.scala

```diff
@@ -238,73 +241,84 @@ object GnirsControllerEpics {
 
     val (lowNoise, digitalAvgs) = readModeEncoder.encode(config.readMode)
 
-    val expTimeWriter = smartSetDoubleParam(expTimeTolerance)(config.exposureTime.toSeconds,
+    val expTimeWriter = smartSetDoubleParamF(expTimeTolerance)(config.exposureTime.toSeconds,
       epicsSys.exposureTime, dcCmd.setExposureTime(config.exposureTime.toSeconds))
 
-    val coaddsWriter = smartSetParam(config.coadds, epicsSys.numCoadds,
+    val coaddsWriter = smartSetParamF(config.coadds, epicsSys.numCoadds,
       dcCmd.setCoadds(config.coadds))
 
     // Value read from the instrument is the negative of what was set
-    val biasWriter = smartSetDoubleParam(biasTolerance)(-encode(config.wellDepth), epicsSys.detBias,
+    val biasWriter = smartSetDoubleParamF(biasTolerance)(-encode(config.wellDepth), epicsSys.detBias,
       dcCmd.setDetBias(encode(config.wellDepth)))
 
-    val lowNoiseWriter = smartSetParam(lowNoise, epicsSys.lowNoise, dcCmd.setLowNoise(lowNoise))
+    val lowNoiseWriter = smartSetParamF(lowNoise, epicsSys.lowNoise, dcCmd.setLowNoise(lowNoise))
 
-    val digitalAvgsWriter = smartSetParam(digitalAvgs, epicsSys.digitalAvgs,
+    val digitalAvgsWriter = smartSetParamF(digitalAvgs, epicsSys.digitalAvgs,
       dcCmd.setDigitalAvgs(digitalAvgs))
 
-    val params =  expTimeWriter ++ coaddsWriter ++ biasWriter ++ lowNoiseWriter ++ digitalAvgsWriter
+    val params = List(expTimeWriter, coaddsWriter, biasWriter, lowNoiseWriter, digitalAvgsWriter)
 
-    if(params.isEmpty) SeqAction(EpicsCommand.Completed)
-    else params.sequence. void*>
-      dcCmd.setTimeout(DefaultTimeout) *>
-      dcCmd.post
+    executeIfNeeded(params,
+      dcCmd.setTimeout[IO](DefaultTimeout) *>
+      dcCmd.post[IO])
   }
 
-  def apply(): GnirsController[IO] =
+  val checkDhs =
+    failUnlessM(
+      epicsSys.dhsConnected.map(_.forall(identity)),
+                SeqexecFailure.Execution("GNIRS is not connected to DHS"))
+
+  val checkArray =
+    failUnlessM(
+      epicsSys.arrayActive.map(_.forall(identity)),
```

Comment:
@jluhrs Previously `arrayActive` was checked with `exists`. Was there a reason for that?

Reply:
It was `GnirsEpics.instance.arrayActive.exists(identity)`, meaning that to pass, the value must be read properly and must be true. (i.e. it must be `Some(true)`). Now `arrayActive` gives a `F[Option[Boolean]]`. `forall` is true for `None`, therefore your code will not stop the observation if it cannot verify the state of the array. That is wrong.

Reply:
It seems we should we apply the same logic for the DHS checks, WDYT?

Reply:
Yes. GNIRS can crash if you try to take an image when it is not connected to the DHS or the detector array is not active.